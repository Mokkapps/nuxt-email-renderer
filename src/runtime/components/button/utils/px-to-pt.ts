export const pxToPt = (px: number): number | null =>
  !Number.isNaN(Number(px)) ? (px * 3) / 4 : null
