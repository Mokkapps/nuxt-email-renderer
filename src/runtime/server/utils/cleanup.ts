export function cleanup(str: string) {
  if (!str)
    return str

  return str
    .replace(/ data-v-inspector="[^"]*"/g, '')
    .replace(/<!--\[-->/g, '')
    .replace(/<!--\]-->/g, '')
    .replace(/<!--[\s\S]*?-->/g, '')
    // Only remove style tags that don't contain @font-face (preserve EFont styles)
    .replace(/<style>(?![\s\S]*@font-face)[\s\S]*?<\/style>/g, '')
    .replace(/<script>[\s\S]*?<\/script>/g, '')
    .replace(/<script[^>]*>[\s\S]*?<\/script>/g, '')
    .replace(/<template>/g, '')
    .replace(/<template[^>]*>/g, '')
    .replace(/<\/template>/g, '')
    .replace(/<\/tailwind-clean-component>/g, '')
}
