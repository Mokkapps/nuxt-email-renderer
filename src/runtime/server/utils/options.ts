import type { HtmlToTextOptions } from 'html-to-text'

export interface I18nRenderOptions {
  /**
   * Enable i18n for this specific render
   */
  enabled: boolean
  /**
   * The locale to use for this render
   */
  locale?: string
  /**
   * Default/fallback locale
   */
  defaultLocale?: string
  /**
   * Use an existing i18n instance from the app
   */
  useAppInstance?: boolean
  /**
   * The existing i18n instance (if useAppInstance is true)
   */
  appInstance?: any
  /**
   * Directory containing locale message files
   */
  localesDir?: string
  /**
   * Pre-loaded locale messages (alternative to localesDir)
   */
  messages?: Record<string, any>
}

export type Options = {
  pretty?: boolean
  /**
   * i18n configuration for this render
   */
  i18n?: I18nRenderOptions
} & (
  | {
    plainText?: false
  }
  | {
    plainText?: true
    /**
     * These are options you can pass down directly to the library we use for
     * converting the rendered email's HTML into plain text.
     *
     * @see https://github.com/html-to-text/node-html-to-text
     */
    htmlToTextOptions?: HtmlToTextOptions
  }
  )
