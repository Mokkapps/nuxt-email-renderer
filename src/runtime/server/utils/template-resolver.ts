import type { Component } from 'vue'
import type { EmailTemplateInfo } from './virtual-templates'

// Email templates and their metadata loaded from the virtual module
let emailTemplates: Record<string, Component> = {}
let emailTemplateMapping: Record<string, EmailTemplateInfo> = {}

// Initialization state tracking
let initialized = false
let initPromise: Promise<void> | null = null

/**
 * Initializes email templates by importing the virtual module
 * The virtual module is generated by Nitro during build/dev startup
 */
async function initializeTemplates() {
  // Prevent multiple concurrent initializations
  if (initPromise) {
    return initPromise
  }

  initPromise = (async () => {
    try {
      // @ts-expect-error - #email-templates is a Nitro virtual module generated at build time
      const virtualModule = await import('#email-templates')

      emailTemplates = virtualModule.emailTemplates || {}
      emailTemplateMapping = virtualModule.emailTemplateMapping || {}

      initialized = true
    }
    catch {
      // Silently handle missing virtual module - this is expected when:
      // 1. The consuming app doesn't have any email templates
      // 2. During prerendering when the module hasn't been properly initialized
      // 3. In environments where email rendering is not needed

      // Initialize with empty objects if module is not available
      emailTemplates = {}
      emailTemplateMapping = {}
      initialized = true // Mark as initialized to prevent retries
    }
    finally {
      initPromise = null
    }
  })()

  return initPromise
}

/**
 * Ensures templates are initialized before use
 */
async function ensureInitialized() {
  if (!initialized) {
    await initializeTemplates()
  }
}

/**
 * Get an email template component by name
 */
export async function getEmailTemplate(
  templateName: string,
): Promise<Component | null> {
  await ensureInitialized()

  const template = emailTemplates[templateName]

  if (!template) {
    console.warn(`[nuxt-email-renderer] Template "${templateName}" not found`)
    return null
  }

  return template
}

/**
 * Get template metadata by name
 */
export async function getEmailTemplateInfo(
  templateName: string,
): Promise<EmailTemplateInfo | null> {
  await ensureInitialized()

  return emailTemplateMapping[templateName] || null
}

/**
 * Get all available email templates
 */
export async function getAllEmailTemplates(): Promise<EmailTemplateInfo[]> {
  await ensureInitialized()

  return Object.values(emailTemplateMapping)
}

/**
 * Check if a template exists
 */
export async function hasEmailTemplate(templateName: string): Promise<boolean> {
  await ensureInitialized()

  return templateName in emailTemplates
}
