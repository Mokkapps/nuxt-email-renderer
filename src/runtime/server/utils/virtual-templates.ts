import { readdir, stat } from 'node:fs/promises'
import { join, basename, extname } from 'node:path'
import { existsSync } from 'node:fs'

export interface EmailTemplateInfo {
  name: string
  filename: string
  displayName: string
  importPath: string
  filePath: string
}

export interface EmailTemplateMapping {
  [templateName: string]: EmailTemplateInfo
}

/**
 * Scans the emails directory and generates template mappings
 */
export async function generateTemplateMapping(
  emailsDir: string,
): Promise<EmailTemplateMapping> {
  const mapping: EmailTemplateMapping = {}

  if (!existsSync(emailsDir)) {
    console.warn(`Email templates directory not found: ${emailsDir}`)
    return mapping
  }

  try {
    const files = await readdir(emailsDir)

    for (const file of files) {
      if (!file.endsWith('.vue')) continue

      const filePath = join(emailsDir, file)
      const fileStats = await stat(filePath)

      if (!fileStats.isFile()) continue

      const name = basename(file, extname(file))
      const displayName = name.replace(/([A-Z])/g, ' $1').trim()

      // Create a virtual import path that Nitro can resolve
      const importPath = `#email-templates/${name}`

      mapping[name] = {
        name,
        filename: file,
        displayName,
        importPath,
        filePath,
      }
    }
  }
  catch (error) {
    console.error('Error scanning email templates directory:', error)
  }

  return mapping
}

/**
 * Generates virtual module content for email templates
 */
export function generateVirtualModule(mapping: EmailTemplateMapping): string {
  const imports: string[] = []
  const exports: string[] = []
  const mappingEntries: string[] = []

  // Helper to safely escape strings for JavaScript
  const escapeString = (str: string) => JSON.stringify(str)

  Object.values(mapping).forEach((template, index) => {
    const importName = `Template${index}`
    imports.push(
      `import ${importName} from ${escapeString(template.filePath)}`,
    )
    exports.push(`  ${escapeString(template.name)}: ${importName}`)

    // Generate mapping entry as valid JavaScript object
    mappingEntries.push(`  ${escapeString(template.name)}: {
    name: ${escapeString(template.name)},
    filename: ${escapeString(template.filename)},
    displayName: ${escapeString(template.displayName)},
    importPath: ${escapeString(template.importPath)},
    filePath: ${escapeString(template.filePath)}
  }`)
  })

  // Handle empty mapping case
  const templatesContent = exports.length > 0 ? exports.join(',\n') : ''
  const mappingContent
    = mappingEntries.length > 0 ? mappingEntries.join(',\n') : ''

  return `${imports.join('\n')}

export const emailTemplates = {
${templatesContent}
}

export const emailTemplateMapping = {
${mappingContent}
}

export function getEmailTemplate(name) {
  return emailTemplates[name]
}
`
}
