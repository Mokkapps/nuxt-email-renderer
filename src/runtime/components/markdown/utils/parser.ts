import type { Renderer } from 'marked'
import { marked } from 'marked'
import type { StylesType } from './types'
import { initRenderer } from './utils'

export class MarkdownParser {
  private readonly renderer: Renderer

  constructor({ customStyles }: { customStyles?: StylesType }) {
    this.renderer = initRenderer({ customStyles })
  }

  parse(markdown: string) {
    return marked.parse(markdown, { renderer: this.renderer })
  }
}
