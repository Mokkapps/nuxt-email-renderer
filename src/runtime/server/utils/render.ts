import { renderToString } from 'vue/server-renderer'
import type { AllowedComponentProps, Component, VNodeProps } from 'vue'
import { createSSRApp } from 'vue'
import type { Options } from './options'
import { convert } from 'html-to-text'
import { pretty } from './pretty'
import { plainTextSelectors } from './plainTextSelectors'
import { cleanup } from './cleanup'

// Import all components from the index file
import { emailComponents } from '../../components/index'

async function registerEmailComponents(app: ReturnType<typeof createSSRApp>) {
  // Register all email components automatically
  for (const [name, componentImporter] of Object.entries(emailComponents)) {
    const component = await componentImporter()
    app.component(name, component.default)
  }
}

export type ExtractComponentProps<TComponent>
  = TComponent extends new () => {
    $props: infer P
  }
    ? Omit<P, keyof VNodeProps | keyof AllowedComponentProps>
    : never

export interface VNode {
  type: string
  props: {
    style?: Record<string, string | number>
    children?: string | VNode | VNode[]
    [prop: string]: unknown
  }
}

export async function render<T extends Component>(
  component: T,
  props?: ExtractComponentProps<T>,
  options?: Options,
) {
  const doctype = '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">'
  const App = createSSRApp(component, props || {})

  // Register all email components automatically
  await registerEmailComponents(App)

  const markup = await renderToString(App)
  if (options?.plainText) {
    return convert(markup, {
      selectors: plainTextSelectors,
      ...(options?.plainText === true ? options.htmlToTextOptions : {}),
    })
  }

  const doc = `${doctype}${cleanup(markup)}`

  if (options && options.pretty) {
    return pretty(doc)
  }

  return doc
}
