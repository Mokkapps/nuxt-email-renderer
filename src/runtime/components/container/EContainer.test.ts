import { describe, expect, it } from 'vitest'
import { h } from 'vue'
import { render } from '../../server/utils/render'
import { EContainer } from './EContainer.vue'

describe('<EContainer> component', () => {
  it('renders children correctly', async () => {
    const testMessage = 'Test message'
    const component = h(EContainer, [
      testMessage,
    ])
    const html = await render(component)
    expect(html).toContain(testMessage)
  })

  it('passes style and other props correctly', async () => {
    const style = { backgroundColor: 'red' }

    const component = {
      render() {
        return h(EContainer, { 'data-testid': 'container-test', style }, {
          default: () => 'TEST',
        })
      },
    }
    const html = await render(component)

    expect(html).toContain('style="max-width:37.5em;background-color:red;"')
    expect(html).toContain('data-testid="container-test"')
  })

  it('renders correctly', async () => {
    const component = {
      render() {
        return h(EContainer, {}, {
          default: () => 'Lorem ipsum',
        })
      },
    }
    const html = await render(component)
    expect(html).toMatchSnapshot()
  })
})
