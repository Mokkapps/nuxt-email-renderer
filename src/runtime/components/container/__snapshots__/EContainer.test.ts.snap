// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<EContainer> component > renders correctly 1`] = `"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd"><table align="center" width="100%" role="presentation" cellspacing="0" cellpadding="0" border="0" style="max-width:37.5em;"><tbody><tr style="width: 100%"><td>Lorem ipsum</td></tr></tbody></table>"`;
